src/simsearch.h

```diff
@@ -162,7 +184,67 @@ namespace dd
     bool _built_index = false; /**< whether the index has been built. */
     bool _map_populate = true; /**< whether to use MAP_POPULATE when mmapping the full index. */
   };
-  
+#endif
+
+#ifdef USE_FAISS
+  class FaissSE
+  {
+  public:
+    FaissSE(const int &f, const std::string &model_repo);
+    ~FaissSE();
+
+    // interface
+    void create_index();
+
+    void update_index();
+
+    void remove_index();
+
+    void index(const URIData &uri,
+               const std::vector<double> &data);
+
+    void index(const std::vector<URIData> &uris,
+               const std::vector<std::vector<double>> &datas);
+
+    void search(const std::vector<double> &vec,
+                const int &nn,
+                std::vector<URIData> &uris,
+                std::vector<double> &distances);
+
+
+    void train();
+    void add_to_db(const int &idx, const URIData &fmap);
+    void get_from_db(const int &idx, URIData &fmap);
+
+    faiss::Index * _findex = nullptr;
+    std::string _index_key;
+
+    int _f = 128; /**< indexed vector length. */
+    long int _index_size = 0;
+    std::string _model_repo; /**< model directory */
+    const std::string _db_name = "names.bin";
+    const std::string _db_backend = "lmdb";
+    const std::string _index_name = "index.faiss";
+    const std::string _il_name = "index_mmap.faiss";
+    caffe::db::DB *_db = nullptr;
+    std::unique_ptr<caffe::db::Transaction> _txn;
+    int _count_put = 0;
+    int _count_put_max = 1000;
+    int _train_samples_size = 100000;
+    bool _ondisk = true;
+
+    std::vector<float> _train_samples;
+
+#ifdef USE_GPU_FAISS
+    bool _gpu = false;
+    faiss::Index * _gpu_index;
+    std::vector<faiss::gpu::GpuResources*> _gpu_res;
+    std::vector<int> _gpuids;
+    int _nprobe = -1;
```

Comment:
Just FYI - I tried building with `-DUSE_CPU_ONLY=ON` and got a compile-time error that `_nprobe` was not a member of `FaissSE`.
```
In file included from /opt/deepdetect/src/mlservice.h:26:0,
                 from /opt/deepdetect/src/services.h:26,
                 from /opt/deepdetect/src/apistrategy.h:26,
                 from /opt/deepdetect/src/deepdetect.h:25,
                 from /opt/deepdetect/src/deepdetect.cc:22:
/opt/deepdetect/src/mlmodel.h: In member function 'void dd::MLModel::create_sim_search(const int&, dd::APIData)':
/opt/deepdetect/src/mlmodel.h:126:20: error: 'class dd::FaissSE' has no member named '_nprobe'
         _se->_tse->_nprobe = ad.get("nprobe").get<int>();
                    ^
In file included from /opt/deepdetect/src/outputconnectorstrategy.h:116:0,
                 from /opt/deepdetect/src/mlservice.h:27,
                 from /opt/deepdetect/src/services.h:26,
                 from /opt/deepdetect/src/apistrategy.h:26,
                 from /opt/deepdetect/src/deepdetect.h:25,
                 from /opt/deepdetect/src/deepdetect.cc:22:
/opt/deepdetect/src/supervisedoutputconnector.h: In member function 'void dd::SupervisedOutput::finalize(const dd::APIData&, dd::APIData&, dd::MLModel*)':
/opt/deepdetect/src/supervisedoutputconnector.h:521:25: error: 'class dd::FaissSE' has no member named '_nprobe'
         mlm->_se->_tse->_nprobe = ad_in.get("nprobe").get<int>();
                         ^
In file included from /opt/deepdetect/src/outputconnectorstrategy.h:117:0,
                 from /opt/deepdetect/src/mlservice.h:27,
                 from /opt/deepdetect/src/services.h:26,
                 from /opt/deepdetect/src/apistrategy.h:26,
                 from /opt/deepdetect/src/deepdetect.h:25,
                 from /opt/deepdetect/src/deepdetect.cc:22:
/opt/deepdetect/src/unsupervisedoutputconnector.h: In member function 'void dd::UnsupervisedOutput::finalize(const dd::APIData&, dd::APIData&, dd::MLModel*)':
/opt/deepdetect/src/unsupervisedoutputconnector.h:222:25: error: 'class dd::FaissSE' has no member named '_nprobe'
         mlm->_se->_tse->_nprobe = ad_in.get("nprobe").get<int>();
```

Reply:
that should be fixed, thanks a lot !